swap/swap.go

```diff
@@ -1,98 +1,451 @@
+func (s *Swap) sendCheque(peer enode.ID) error {
+	swapPeer, err := s.getPeer(peer)
+	if err != nil {
+		return fmt.Errorf("error while getting peer: %s", err.Error())
+	}
+	cheque, err := s.createCheque(peer)
+	if err != nil {
+		return fmt.Errorf("error while creating cheque: %s", err.Error())
+	}
+
+	log.Info("sending cheque", "serial", cheque.ChequeParams.Serial, "amount", cheque.ChequeParams.Amount, "beneficiary", cheque.Beneficiary, "contract", cheque.Contract)
+	s.cheques[peer] = cheque
+
+	err = s.store.Put(sentChequeKey(peer), &cheque)
+	if err != nil {
+		return fmt.Errorf("error while storing the last cheque: %s", err.Error())
+	}
+
+	emit := &EmitChequeMsg{
+		Cheque: cheque,
+	}
+
+	// reset balance;
+	err = s.resetBalance(peer, int64(cheque.Amount))
+	if err != nil {
+		return err
 	}
-	return 0, errors.New("Peer not found")
+
+	return swapPeer.Send(context.Background(), emit)
 }
 
-//load balances from the state store (persisted)
-func (s *Swap) loadState(peer *protocols.Peer) (err error) {
-	var peerBalance int64
-	peerID := peer.ID()
+// Create a Cheque structure emitted to a specific peer as a beneficiary
+// The serial and amount of the cheque will depend on the last cheque and current balance for this peer
+// The cheque will be signed and point to the issuer's contract
+func (s *Swap) createCheque(peer enode.ID) (*Cheque, error) {
+	var cheque *Cheque
+	var err error
+
+	swapPeer, err := s.getPeer(peer)
+	if err != nil {
+		return nil, fmt.Errorf("error while getting peer: %s", err.Error())
+	}
+	beneficiary := swapPeer.beneficiary
+
+	peerBalance := s.balances[peer]
+	// the balance should be negative here, we take the absolute value:
+	honey := uint64(-peerBalance)
+
+	// convert honey to ETH
+	var amount uint64
+	amount, err = s.oracle.GetPrice(honey)
+	if err != nil {
+		return nil, fmt.Errorf("error getting price from oracle: %s", err.Error())
+	}
+
+	// we need to ignore the error check when loading from the StateStore,
+	// as an error might indicate that there is no existing cheque, which
+	// could mean it's the first interaction, which is absolutely valid
+	err = s.loadLastSentCheque(peer)
+	if err != nil && err != state.ErrNotFound {
+		return nil, err
+	}
+	lastCheque := s.cheques[peer]
+
+	serial := uint64(1)
+	if lastCheque != nil {
+		cheque = &Cheque{
+			ChequeParams: ChequeParams{
+				Serial: lastCheque.Serial + serial,
+				Amount: lastCheque.Amount + amount,
+			},
+		}
+	} else {
+		cheque = &Cheque{
+			ChequeParams: ChequeParams{
+				Serial: serial,
+				Amount: amount,
+			},
+		}
+	}
+	cheque.ChequeParams.Timeout = defaultCashInDelay
+	cheque.ChequeParams.Contract = s.owner.Contract
+	cheque.ChequeParams.Honey = honey
+	cheque.Beneficiary = beneficiary
+
+	cheque.Signature, err = s.signContent(cheque)
+
+	return cheque, err
+}
+
+// Balance returns the balance for a given peer
+func (s *Swap) Balance(peer enode.ID) (int64, error) {
+	var err error
+	// check the balance in memory
```

Comment:
comments not necessary, code self-explanatory

Reply:
fixed